### DISCUSSION TITLE
gridRef not populating in AgGridSolid example

### DISCUSSION THREAD

**Discussion by klmcwhirter** [solidjs/solid#1354] [Q&A] [answered]
In the example given on [https://www.ag-grid.com/react-data-grid/solidjs/#grid-api](https://www.ag-grid.com/react-data-grid/solidjs/#grid-api) the gridRef variable is not ever set. _I am aware of the typo in that example._

If I use a normal HTML element - say ```<div ref={divRef}>...</div>``` that seems to work fine.

Note I did encounter the discussion in #116 by Ryan about ref lifecycle management - [https://github.com/solidjs/solid/issues/116](https://github.com/solidjs/solid/issues/116).

Neither ref or forwardRef seems to work (with or without the setTimeout), it is still undefined in an onMount callback as well.

As you can tell I am in the early stages of the learning curve. I have spent the last 7 years or so with Angular and have no React experience. And while I did play around with knockoutjs years ago ;), for all intents and purposes SolidJS is my first reactive framework.

I did get it to work using a signal and the callback approach ...
```
...
const [gridRef, setGridRef] = createSignal<AgGridSolidRef>(undefined);
...
<AgGridSolid
  ref={ele => setGridRef(ele)}
  onFilterChanged={() => gridRef().api. /* use api */}
  onGridReady={() => gridRef().api. /* use api */}
  ...
  />
```

But ...
* am I doing something wrong? I feel like I am missing some understanding of how the gridRef variable gets set looking at both the ag-grid example code and the code given at [ref](https://www.solidjs.com/docs/latest/api#ref) in the api docs - myDiv there. It seems like magic to me. Does the TSX compiler do that?
* am I encountering a bug in the SolidJS TSX compiler? Yes, I am using TypeScript if that matters.
* is this an issue with the AgGridSolid implementation?
* or do I really need to use a signal for some reason - even though Ryan discourages the usage of a signal for this use case in some post I encountered somewhere that linked to #116 ?

Thoughts?

Thanks.

**Comment by ryansolid** [comment, marked as answer]
Yeah that discussion is outdated. We got rid of `forwardRef` shortly after that.

What was posted looks like it should work (other than the typo). A component ref compiles to:

```js
let refEl;

return _$createComponent(Component, {
    ref(r$) {
      const _ref$ = refEl;
      typeof _ref$ === "function" ? _ref$(r$) : refEl = r$;
    },
    ...
  });
```

**Reply by klmcwhirter** [reply]
That is what I thought too - comparing to your official docs. Let me put together a working / not-working example set to illustrate.

Busy week on the day job this week - I'll get something out Saturday.

Thanks.

**Comment by klmcwhirter** [comment]
OK, it does seem to be working if I click on the buttons, but the console.log messages from onMount and createEffect don't show that.

But please do tell if this (the signal approach in place in aggrid.tsx) is what you had in mind or not.

_Please note I still have not chased down how to do routing in solid yet._
[EDIT] routing was straightforward - added it

Please look at aggrid.tsx (uses a signal for gridRef) and aggrid2.tsx (follows the examples verbatim).

[example repo link](https://github.com/klmcwhirter/explore-aggrid-solid-tw/tree/master/src)

Thanks for the help.
-k

**Comment by klmcwhirter** [comment]
Ryan, BTW - all those years ago when I was playing with knockoutjs it was working with IdeaBlade's DevForce with help from Ward Bell.

Ironic that you have taken inspiration from knockoutjs.

Great framework! I wish I had this available before fighting all these years with angularjs and then Angular.

Such is life ...

But, great job - both with the framework and the adoption plan.

I was really surprised to learn about AG Grid support.

You are a genius in my book ...

-k
